streamer/src/bounded_streamer.rs

```diff
@@ -0,0 +1,440 @@
+        }
+        Self {
+            signal_sender: self.signal_sender.clone(),
+            data: self.data.clone(),
+        }
+    }
+}
+
+impl Drop for BoundedPacketBatchSender {
+    fn drop(&mut self) {
+        let dropping_last_sender = self.data.sender_count.fetch_sub(1, Ordering::SeqCst) == 1;
+        if dropping_last_sender {
+            // notify receivers, otherwise they may be waiting forever on a
+            // disconnected channel
+            self.signal_sender.try_send(()).unwrap_or(());
+        }
+    }
+}
+
+impl PacketBatchChannelData {
+    fn add_packet_count(&self, amount: usize) {
+        self.packet_count.fetch_add(amount, Ordering::Relaxed);
+    }
+    fn sub_packet_count(&self, amount: usize) {
+        self.packet_count.fetch_sub(amount, Ordering::Relaxed);
+    }
+}
+
+enum TryRecvError {
+    NoData,
+    DisconnectedAndNoData,
+}
+
+impl BoundedPacketBatchReceiver {
+    /// Receives up to around `packet_count_hint` packets from the channel.
+    ///
+    /// If no data is available, the function will wait up to `timeout`, then
+    /// return RecvTimeoutError::Timeout if no data came in.
+    ///
+    /// If all senders have been dropped and there's no data avilable, it returns
+    /// RecvTimeoutError::Disconnected.
+    ///
+    /// Returns (list of batches, packet count)
+    pub fn recv_timeout(
+        &self,
+        packet_count_hint: usize,
+        timeout: Duration,
+    ) -> Result<(Vec<PacketBatch>, usize), RecvTimeoutError> {
+        let deadline = Instant::now() + timeout;
+        loop {
+            self.signal_receiver.recv_deadline(deadline)?;
+            return match self.try_recv(packet_count_hint) {
+                Ok(r) => Ok(r),
+                Err(TryRecvError::NoData) => continue,
+                Err(TryRecvError::DisconnectedAndNoData) => Err(RecvTimeoutError::Disconnected),
+            };
+        }
+    }
+
+    /// Receives up to around `packet_count_hint` packets from the channel.
+    ///
+    /// Waits until there's data or the channel has been disconnected.
+    ///
+    /// Returns (vec-of-batches, packet count)
+    pub fn recv(&self, packet_count_hint: usize) -> Result<(Vec<PacketBatch>, usize), RecvError> {
+        loop {
+            self.signal_receiver.recv()?;
+            return match self.try_recv(packet_count_hint) {
+                Ok(r) => Ok(r),
+                Err(TryRecvError::NoData) => continue,
+                Err(TryRecvError::DisconnectedAndNoData) => Err(RecvError),
+            };
+        }
+    }
+
+    // Returns (vec-of-batches, packet-count)
+    fn try_recv(
+        &self,
+        packet_count_hint: usize,
+    ) -> Result<(Vec<PacketBatch>, usize), TryRecvError> {
+        let mut packets = 0;
+        let disconnected = self.data.sender_count.load(Ordering::SeqCst) == 0;
+
+        let mut batches = Vec::with_capacity(min(
+            packet_count_hint / PACKETS_PER_BATCH,
```

Comment:
Is this too optimistic? What about the case where we have very few packets in each batch?

Reply:
It's a new optimization: the old `recv_packet_batches()` didn't control the result vec capacity at all. So it should be an improvement even if this lower bound often underestimates the actual final capacity.

We could add a `* 2`. If 100k packets are requested, that'd shift the min from 781 to 1562 and probably cost around an extra 32kb of memory which may or may not be needed. I'm ambivalent and would probably keep as-as, but increasing this a bit is fine too.

If we wanted to go all-in on optimization, we'd receive into the same Vec every time to skip this reallocation in most cases, but I think the api-niceness cost is not worth the gain.